### PR TITLE
[JENKINS-48638] Deprecate Jenkins.getInstance in favor of (usually) get

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-48638](https://issues.jenkins-ci.org/browse/JENKINS-48638)

Essentially the opposite of #3192: safer and (IMO) much clearer, though at the expense of introducing lots of deprecation warnings into plugins updating past this baseline. Plugins can switch to `getActiveInstance` today, fixing FindBugs, at the expense of suffering deprecation warnings both before and after the baseline. Pick your poison.

### Proposed changelog entries

* `Jenkins.getInstance` is now deprecated as its semantics have been a source of confusion for some time. Use `get` in typical cases.

@jenkinsci/code-reviewers

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
b1f

**Comment by daniel-beck** [comment]
🤦‍♂️ No please no it hurts make it stop 😭

**Review by oleg-nenashev — DISMISSED** [review]
Obviously 🐝 from me

**Review by stephenc — DISMISSED** [review]
Please no

**Comment by stephenc** [comment]
If we want a cleaner version, move to revert https://github.com/jenkinsci/jenkins/pull/2297 and restore `getInstance()` as `Nullable`. The reality is that basically any plugin code can assume `getInstance()` is non-null... unless you are doing strange things with background threads (and violating the servlet spec in the process)

Switching to `getActiveInstance()` *again* will just make a mockery of the project IMHO.

Perhaps the simplest thing is to just turn off findbugs for `getInstance()` and claim it is always non-null and be done

**Comment by jglick** [comment]
> restore `getInstance()` as `Nullable`

This would again encourage plugin code to assume it is nonnull and then get cryptic NPEs when in fact things _are_ written strangely; and does little to clean up confusion about the actual nullability of this method. I think it is better for plugin authors to consciously decide what they intend the semantics to be.

> just turn off findbugs for `getInstance()`

This patch accomplishes that already. Existing code written either before or after the mess you made will pass FindBugs nullability checks, so we no longer need to waste time rewriting code every time we update parent POMs. Whenever a plugin author is ready to pay attention to deprecation warnings, they can take a few seconds to make a conscious decision to switch to `getActiveInstance` (usually) and delete any associated null checks.

**Comment by stephenc** [comment]
For plugins.... except crazy pipeline... and people incorrectly using it from agents... it should be non-null which is why `@Nullable` was the correct choice if it can return `null`.

The promise was that we could start throwing ISEs when we went from 1.x to 2.0.

If you want to start throwing blame around, the original change to `getActiveInstance()` was the stupidest thing ever. The amount of code changes just to handle the change from `getInstance()` to `getActiveInstance()`... never mind that it is longer to type...

If you were going to get a NPE when calling `getInstance()` then you were also going to get an `ISE`.

I have a solution.

Change `getInstance()` to throw an `NPE` if there is no instance... but at least include a message saying why it is null. You have fixed subversion, we are so close to undoing the original madness of making `getInstance()` be `@CheckForNull`

This PR is, IMHO, completely the wrong approach

**Comment by oleg-nenashev** [comment]
> Change getInstance() to throw an NPE if there is no instance...

You remember what happened last time we did that... I do not feel we want to try again without a serious preparation to that. 

I agree with @stephenc that going back and forth on `getActiveInstance()` is not good from the community perspective. Just to provide the full track of changes by me, @stephenc and @batmat 

* 2014 - @oleg-nenashev introduced `getActiveInstance()`, `getInstanceOrNull()` and made `getInstance()` `@CheckForNull`
* Mar 2016 - @stephenc Introduced https://github.com/jenkinsci/jenkins/pull/2094, which introduced `getInstanceOrNull()`, deprecated `getActiveInstance()` and made `getInstance()` nullable. 
* May 2016 - @batmat made the `getInstance()` Nonnull in https://github.com/jenkinsci/jenkins/pull/2297. 
* May 2016 - Then we had a regression, and the change partially reverted in https://github.com/jenkinsci/jenkins/pull/2355. `getInstance()` was able to return `null` since that (FindBugs suppressed)

I am a strong supporter of `getInstanceOrNull()`/`getActiveInstance()` approach (and hence or this PR), but from the community perspective I feel that we do not want plugin maintainers to rewrite all their code again and again.

My proposals: 

1) Keep it as is for now and let it miss one LTS baseline
2) Discuss the longer-term plan at the FOSDEM contributor summit (API Deprecation policy, other architecture changes in Jenkins)
3) Depending on the decisions there, decide how we proceed with core annotations and APIs. JEP it if needed

**Comment by jglick** [comment]
> The amount of code changes just to handle the change from `getInstance()` to `getActiveInstance()`

Unlike some earlier changes, there is no timeline forced upon plugin developers by FindBugs errors. I would not object to removing the deprecation from `getInstance()` if that is what it takes to get this in. I would still advise people to use one of the more explicit variants.

> never mind that it is longer to type

We could introduce a

```java
public static @Nonnull Jenkins get() {
    return getActiveInstance();
}
```

**Comment by stephenc** [comment]
I think either `Jenkins.instance()` or `Jenkins.get()` should be the recommendation. There is no reason to access the Singleton from jelly (given that it is already bound to `app`) so no need to use a JavaBeans style getter (not that java beans means anything on a static field if I recall the spec correctly)

Then either deprecate everything else or leave them undeprecated.

**Comment by jglick** [comment]
> not that java beans means anything on a static field

Not according to JavaBeans; IIRC Jelly _does_ let you use abbreviated syntax for `static` “getters” (invoked on an instance variable), but anyway as you say `app` suffices in this case.

**Review by stephenc — APPROVED** [review]

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -739,56 +738,57 @@ protected File getRootDirFor(String name) {
...
+     * for example, gather any information you need on the master side before constructing the callable.
+     * If you must do a runtime check whether you are in the master or agent, use {@link JenkinsJVM} rather than this method,
+     * as merely loading the {@link Jenkins} class file into an agent JVM can cause linkage errors under some conditions.
+     * @return The instance. Null if the {@link Jenkins} service has not been started, or was already shut down,
+     *         or we are running on an unrelated JVM, typically an agent.
      * @since 1.653
      */
+    @CLIResolver
     @CheckForNull
     public static Jenkins getInstanceOrNull() {
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: Maybe even `tryGet()` if we want to simplify all methods.

**Review by oleg-nenashev — APPROVED** [review]

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (batmat) [inline core/src/main/java/jenkins/model/Jenkins.java]: Should really be `getOrNull()` then for consistency and conciseness.

**Review by batmat — APPROVED** [review]
`getInstanceOrNull` ought to be `getOrNull` too for consistency IMO.

:bee:

File: core/src/main/java/hudson/model/Hudson.java [hunk 2]
```diff
@@ -70,10 +69,10 @@
     @Deprecated
     private transient final CopyOnWriteList<ComputerListener> computerListeners = ExtensionListView.createCopyOnWriteList(ComputerListener.class);
 
-    /** @deprecated Here only for compatibility. Use {@link Jenkins#getInstance} instead. */
+    /** @deprecated Here only for compatibility. Use {@link Jenkins#get} instead. */
     @Deprecated
     @CLIResolver
-    @Nonnull
+    @Nullable
```
Comment (jtnord) [inline core/src/main/java/hudson/model/Hudson.java]: WHy?  Heel no.  there is no JIRA and this will break the build of many plugins when upgrading the jenkins core

**Review by jtnord — DISMISSED** [review]
not on your nelly and certainly not withiout a JIRA

**Comment by kuisathaverat** [comment]
@jglick IMHO it is ok as it now, maybe as suggest @stephenc you could throw a runtime exception with a clear message of what happened if the instance is not initialized. I recently changed the `getInstanceActive` to `getInstance` in the `SAML Plugin` because was deprecated and that change was pretty accurate for me, is more clear, also if the methods are correctly annotated is a developer responsibility to check the findbugs in order to avoid weird exceptions accessing to the Jenkins instance. For me is a clear `NO` and will be a mess because we will have 3 different ways to access to the Jenkins instance depending on the parent pom that you use that it is for sure really error prone, with this change my plugin would get a null when call the `getInstance` method when now this is not possible.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 3]
```diff
@@ -739,56 +738,57 @@ protected File getRootDirFor(String name) {
...
      */
-    @Deprecated
     @Nonnull
-    public static Jenkins getActiveInstance() throws IllegalStateException {
-        Jenkins instance = HOLDER.getInstance();
+    public static Jenkins get() throws IllegalStateException {
+        Jenkins instance = getInstanceOrNull();
         if (instance == null) {
-            throw new IllegalStateException("Jenkins has not been started, or was already shut down");
+            throw new IllegalStateException("Jenkins.instance is missing. Read the documentation of Jenkins.getInstanceOrNull to see what you are doing wrong.");
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: if Jenkins.instance where missing you would blow up with some form of linkage error ->  the instance is not missing it is `null`

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (jglick) [reply core/src/main/java/jenkins/model/Jenkins.java]: Since legitimate calls to this method should be quite rare, I saw no need to introduce yet another alias.

**Comment by jglick** [comment]
> with this change my plugin would get a null when call the `getInstance` method when now this is not possible

@kuisathaverat no, that is not true! `getInstance()` can currently return null—the annotation is wrong.

This PR does not change the behavior of any existing method, so it cannot cause any runtime regressions. It does relax the FindBugs mess, so if updating from 1.x straight to here you would not need to edit every call to `getInstance` just to be FB-clean.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 3 above)
Comment (jglick) [reply core/src/main/java/jenkins/model/Jenkins.java]: I am not sure what the distinction is. If you got this far, obviously `Jenkins.class` has been successfully loaded, and `theInstance == null`.

**Comment by jglick** [comment]
> not on your nelly

Could you be more specific…? :-)

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 3 above)
Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java]: The distinction is massive in terms of clarity

Jenkins.Instance is a field. 
A missing field is saying the field is not there which can only be because it was removed from a class (it is no longer defined in the source file yet you expect it to be there) 
Null is not missing, the field is there it has just not been set to anything other than its default (null) (or has been set back to null for some reason) 

The message should help users (is this something wrong with their job, jenkins...) 
Basically the message should be Jenkins.instance is null (unset)   This is a bug in the plugin. Please read the docume....

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -739,56 +738,57 @@ protected File getRootDirFor(String name) {
...
+     * <p>In certain rare cases you may have code that is intended to run before Jenkins starts or while Jenkins is being shut down.
+     * For those rare cases use this method.
+     * <p>In other cases you may have code that might end up running on a remote JVM and not on the Jenkins master.
+     * For those cases you really should rewrite your code so that when the {@link Callable} is sent over the remoting channel
+     * it can do whatever it needs without ever referring to {@link Jenkins};
+     * for example, gather any information you need on the master side before constructing the callable.
+     * If you must do a runtime check whether you are in the master or agent, use {@link JenkinsJVM} rather than this method,
+     * as merely loading the {@link Jenkins} class file into an agent JVM can cause linkage errors under some conditions.
+     * @return The instance. Null if the {@link Jenkins} service has not been started, or was already shut down,
+     *         or we are running on an unrelated JVM, typically an agent.
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: I do wonder why this doesn't check that it is running on the master and throw an exception if not (as that is a clear violation of the intent of this contract) and from this java client something that should never be done.

**Review by jtnord — COMMENTED** [review]
The JIRA makes the intent here much clearer, thankyou.  I think this could be improved upon, it sounds like if an api call should never be used on an agent then it should throw an exception if it is (there would be a clear violation of the contract). The exception message is not clear where the issue lies (unless you are a developer.  An admin should ideally be able to tell if jenkins config, job / build config or a plugin is broken, but in this case it is not clear to them (even if it is clear to a developer))

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 4 above)
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: I agree a new method for this should have a 

```java
if (!JenkinsJVM.isJenkinsJVM()) {
    throw new IllegalStateException("Jenkins instance must not be accessed outside Jenkins");
}
````

by possibly deprecating it and creating a `getOrNull()` to be consistent with the new methods naming. (?)

Comment (jglick) [reply core/src/main/java/jenkins/model/Jenkins.java]: Does not work I think—`isJenkinsJVM` will return false for the same reasons that `Jenkins.instance == null`. It just avoids loading `Jenkins.class` (and thus `javax.servlet.*`, …) in the process.

Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: Not sure what you mean, but the goal would be to have an exception thrown instead of just returning `null` so that plugins developers don't put an `if` on the nullity of `getInstanceOrNull()` anymore when this is always going to be null on an agent's side?

Comment (jglick) [reply core/src/main/java/jenkins/model/Jenkins.java]: Not sure where you are going with this. What use case?

Normal code can simply use `.get()` and be done.

Comment (stephenc) [reply core/src/main/java/jenkins/model/Jenkins.java]: @batmat well at some point we will actually ban `Jenkins` from the agent classloader, so anyone doing that will just plain blow up there and then. I don't think we need to guard further as in that case you still would not get the exception, just a class not found
